File: images/0009-Win32OLE/ie.pl
use Win32::OLE qw(EVENTS in);
use Win32::OLE::Const ('Microsoft Internet Controls');

$| = 1;
my $ie = Win32::OLE->new('InternetExplorer.Application');
Win32::OLE->WithEvents($ie,"WebBrowserEvents","DWebBrowserEvents2");
$ie->{Visible} = 1;
$ie->GoHome();
Win32::OLE->MessageLoop();
while($ie->ReadyState() != READYSTATE_COMPLETE){
}

$ie->Navigate('http://www.ruby-lang.org/');

Win32::OLE->MessageLoop();
while($ie->ReadyState() != READYSTATE_COMPLETE){
}

my $count = 0;
foreach my $element (in $ie->Document->all){
    $count++;
}
print "complete\n ${count} elements found\n";

package WebBrowserEvents;
sub DownloadComplete {
    my ($obj,@args) = @_;
    print "Download Complete\n";
}
sub NavigateComplete2 {
    my ($obj,@args) = @_;
    Win32::OLE->QuitMessageLoop();
}

File: docs/images/0009-Win32OLE/ie.rb
require 'win32ole'

module InternetExplorer
end

ie = WIN32OLE.new('InternetExplorer.Application')
WIN32OLE.const_load(ie,InternetExplorer)

event = WIN32OLE_EVENT.new(ie,"DWebBrowserEvents2")
event.on_event("DownloadComplete") do 
  puts "Download Complete."
end

ie.Visible = true

ie.GoHome
while ie.ReadyState != InternetExplorer::READYSTATE_COMPLETE
  WIN32OLE_EVENT.message_loop
end

ie.Navigate 'http://www.ruby-lang.org/'
while ie.ReadyState != InternetExplorer::READYSTATE_COMPLETE
  WIN32OLE_EVENT.message_loop
end

count = 0
ie.Document.all.each do 
  count += 1
end
puts "complete\n#{count} elements found"
